tests/phpunit/integration/api/html/beansAddAttribute.php

```diff
@@ -23,49 +23,54 @@
 class Tests_BeansAddAttribute extends HTML_Test_Case {
 
 	/**
-	 * Test beans_add_attribute() should add an empty value when no value is given.
+	 * Test beans_add_attribute() should return register the "add" callback to the given ID.
 	 */
-	public function test_should_add_empty_value_when_no_value_given() {
+	public function test_should_register_the_add_callback_to_given_id() {
+		$instance = beans_add_attribute( 'foo', 'data-test', 'test' );
 
-		foreach ( static::$test_markup as $beans_id => $markup ) {
-			$markup_attributes = isset( $markup['attributes'] ) ? $markup['attributes'] : array();
-			$attributes        = beans_add_attribute( $beans_id, 'data-test', '' );
-			$hook              = $beans_id . '_attributes';
+		$this->assertInstanceOf( \_Beans_Attribute::class, $instance );
+		$this->assertSame( 10, has_filter( 'foo_attributes', array( $instance, 'add' ), 10 ) );
 
-			// Run the tests.
-			$this->assertSame( 10, has_filter( $hook, array( $attributes, 'add' ), 10 ) );
-			$expected              = $markup_attributes;
-			$expected['data-test'] = '';
-			$this->assertSame( $expected, apply_filters( $hook, $markup_attributes ) ); // phpcs:ignore WordPress.NamingConventions.PrefixAllGlobals.NonPrefixedHooknameFound -- The hook's name is in the value.
+		// Clean up.
+		remove_filter( 'foo_attributes', array( $instance, 'add' ) );
+	}
+
+	/**
+	 * Test beans_add_attribute() should add the attribute when it does not exist in the given attributes.
+	 */
+	public function test_should_add_the_attribute_when_does_not_exist() {
+
+		foreach ( static::$test_attributes as $beans_id => $markup ) {
+			$hook     = $beans_id . '_attributes';
+			$instance = beans_add_attribute( $beans_id, 'data-test', 'foo' );
+
+			// Run the full systems test by applying the filter.
```

Comment:
This still feels like a system test (nothing is mocked), rather than an integration test (where just two bits of Beans is used and everything else is mocked). There's nothing wrong with having a different suite of tests to split up integration and system tests.

Reply:
So you are advocating that we create 3 different suites of tests:

1. Unit - where we test without the dependencies via mocking/simulating those dependencies.
2. Integration - where we integrate the dependencies but do not test all the way through. 
3. Systems - where we do a full integration and test all the way through.

So in this case of the "add attribute", using these 3 set suites, we'd have;

1. Unit tests for the "add" and "init" methods.
2. Integration tests to ensure that the proper filter callback was registered, i.e. `has_filter()`.
3. Systems tests to fire the filter and test that we get the right results.

Is that what you are suggesting here?

Reply:
In your opinion, you do not want to check all the way through in an integration test.  For example, for `beans_add_attribute()` the expectation are:

1. It creates an attribute object with the given inputs of ID, the target attribute name, and the value to add.
2. It registers the object's callback for the "add" task.

Those are the _technical_ expectations of the `beans_add_attribute()`.

But then we are leaving out the endpoint.  Right?  When we call `beans_add_attribute` in our code, what we are really expecting is that:

1. We've created an object that holds the targets, i.e. for this ID, I want to target this attribute name and add this value to it.
2. Then when the event fires (which happens in the same API), we expect it to add the value to attribute.

Where do we test that?  We could do it in the other HTML function that actually fires the event and we will indirectly by testing the final HTML that is built.  

But my approach is, go one more step right here where we are focusing on adding the attribute. 
 Make sure at the attribute level that the attribute is added for the given inputs.

Yes, we could create a third suite of tests and label them "systems".  But I think that is too much to maintain.  Why?  We'd be repeating the bulk of this code again, simply to test the final result.

My definition of integration test is full integration, with all dependencies functioning as expected.  A unit test tests a function scenarios without the dependencies.  An integration test integrates the dependencies into the test suite to ensure our target function performs per our expectations for the given scenarios.